docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -8,236 +8,342 @@ title: Authorization
-When authorization is required and not yet proven by the client, servers **MUST** respond
-with _HTTP 401 Unauthorized_.
+MCP servers **MUST** follow [RFC 6750 Section 3.1](https://www.rfc-editor.org/rfc/rfc6750#section-3.1) for
+any authorization failures.
 
-Clients initiate the
-[OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#name-authorization-code-grant)
-authorization flow after receiving the _HTTP 401 Unauthorized_.
+For MCP server capability discovery:
 
-The following demonstrates the basic OAuth 2.1 for public clients using PKCE.
+- MCP clients **MUST** parse the `WWW-Authenticate` response to discover the issuer URL, and construct
+  the metadata endpoint from that issuer identifier.  
+- MCP servers **SHOULD** use a third-party authorization server and include its issuer in
+  error responses.  
+- MCP clients **SHOULD** evaluate or enable the user to evaluate the trustworthiness of identity
+  providers.  
+- Authorization Servers **MUST** support
+  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)
+- MCP servers **MUST** respond to an unauthorized request with a `WWW-Authenticate` header containing
+  `issuer`, `scope`, and `resource` identifiers.
+
+The discovery flow is illustrated below:  
 
 ```mermaid
 sequenceDiagram
-    participant B as User-Agent (Browser)
-    participant C as Client
-    participant M as MCP Server
-
-    C->>M: MCP Request
-    M->>C: HTTP 401 Unauthorized
-    Note over C: Generate code_verifier and code_challenge
-    C->>B: Open browser with authorization URL + code_challenge
-    B->>M: GET /authorize
-    Note over M: User logs in and authorizes
-    M->>B: Redirect to callback URL with auth code
-    B->>C: Callback with authorization code
-    C->>M: Token Request with code + code_verifier
-    M->>C: Access Token (+ Refresh Token)
-    C->>M: MCP Request with Access Token
-    Note over C,M: Begin standard MCP message exchange
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AUTHZS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+    MC->>MC: Construct metadata endpoint from issuer identifier
+
+    Note over MC: Continue with authorization flow
 ```
 
-### 2.3 Server Metadata Discovery
+#### 2.3.1 Server Metadata Discovery Headers
+
+MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
+Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.
+
+For example: `MCP-Protocol-Version: 2024-11-05`
+
+### 2.4 Authorization server integration
+
+#### 2.4.1 Overview
+
+MCP servers **MAY** override endpoints on an as-needed basis if the authorization server requires
+additional gestures that are not captured by the specification.
 
-For server capability discovery:
+MCP servers **MUST NOT** complete PKCE or other public client flows on behalf of the client.
```

Comment:
PKCE Is not a public client flow, it is also used by confidential clients.

What is the purpose of this sentence? It seems like you've seen some behavior you're trying to prohibit, but this doesn't really make any sense.

Reply:
Updated and clarified the wording.